text/0031-corejam.md

```diff
@@ -0,0 +1,636 @@
+
+Several host functions (largely in line with the host functions available to Parachain Validation Function code) are supplied. One addition is:
+
+```rust
+/// Determine the preimage of `hash` utilizing the Relay-chain Storage pallet.
+fn lookup(hash: [u8; 32]) -> Vec<u8>;
+```
+
+Other host functions will allow for the possibility of executing a WebAssembly payload (for example, a Parachain Validation Function) or instantiating and entering a subordinate RISCV VM (for example for Actor Progressions).
+
+When applying `refine` from the client code, we must allow for the possibility that the VM exits unexpectedly or does not end. Validators are always rewarded for computing properly authorized Work Packages, including those which include such broken Work Items. But they must be able to report their broken state into the Relay-chain in order to collect their reward. Thus we define a type `WorkResult`:
+
+```rust
+enum WorkError {
+    Timeout,
+    Panic,
+}
+struct WorkResult {
+    class: WorkClass,
+    item_hash: [u8; 32],
+    result: Result<WorkOutput, WorkError>,
+    weight: Weight,
+}
+fn apply_refine(item: WorkItem) -> WorkResult;
+```
+
+The amount of weight used in executing the `refine` function is noted in the `WorkResult` value, and this is used later in order to help apportion on-chain weight (for the Join-Accumulate process) to the Work Classes whose items appear in the Work Packages.
+
+```rust
+/// Secure refrence to a Work Package.
+struct WorkPackageSpec {
+    /// The hash of the SCALE encoded `EncodedWorkPackage`.
+    hash: WorkPackageHash,
+    /// The erasure root of the SCALE encoded `EncodedWorkPackage`.
+    root: ErasureRoot,
+    /// The length in bytes of SCALE encoded `EncodedWorkPackage`.
+    len: u32,
+}
+/// Execution report of a Work Package, mainly comprising the Results from the Refinement
+/// of its Work Items.
+struct WorkReport {
+    /// The specification of the underlying Work Package.
+    package_id: WorkPackageId,
+    /// The context of the underlying Work Package.
+    context: Context,
+    /// The Core index under which the Work Package was Refined to generate the Report.
+    core_index: CoreIndex,
+    /// The results of the evaluation of the Items in the underlying Work Package.
+    results: BoundedVec<WorkResult, MaxWorkItemsInPackage>,
+}
+/// Multiple signatures are consolidated into a single Attestation in a space-efficient
+/// manner using a `BitVec` to succinctly express which validators have attested.
```

Comment:
Is the data here dense enough to make a `BitVec` more space-efficient?  
With more than 16 cores it could be less efficient than storing `Vec<u16>` (assuming there are no more than 2¹⁶ validators).

Reply:
Yeah; this is copied from the existing impl. Perhaps it can be optimized. Comments @eskimor ?

Reply:
`BitVec` is internally probably also just something like `Vec<u16>`. However, `BitVec` provides you all the nice utility functions that you want to have for bit operations etc.

Reply:
I think the point is that if there's only expected to be 5 signatories but 1000 validators, then a `BitVec` would always encode to 1000 bits, yet a `Vec<u16>` would typically encode to ~80 bits.

Reply:
I assume the `BitVec` here represents indices within a single group (up to 5 bits/1 byte), so it's already as efficient as it gets.

Reply:
That's right. It's the index of the validator within the group - so the encoded size is `sizeof(Compact(1usize)) + 1` for any validator group size smaller than 8. We could get away with a single u16 but `BitVec` is the only way to handle arbitrary validator group sizes.